apb pkg public interface overhaul

eriknelson opened the pull request:
**Describe what this PR does and why we need it**:
CC: @maleck13, this is the intended follow-up for #619. We identified the way we'd like the public interfaces of the apb pkg to move in light of trying to think more consciously about what the experience should be like consuming `libapb`. Namely:

* There should be a suite of methods for callers to run APBs both asynchronously, and synchronously.
* The async calls are non-blocking, and should push updates back to the caller over a `StatusMessage` channel.
* The sync calls should block, and return with a result.

Changes proposed in this pull request:
 - Implement the above items. Introduces an `Executor` type that manages the APB lifecycle, and exposes the various methods for execution.
 - Update the broker calling code to use the new apb pkg correctly.
 - Job code has become virtually identical, so this removes the individual jobs favoring a more generic `apbJob` type, with concrete wrapper methods that call into the generic job.

Couple blockers before a merge:
* It's obviously a pretty big change to how we're doing things, so I would like to get a lot of review attention and testing.
* ~~I deleted a lot of non-relevant tests and haven't added any to test the new code paths, so I think we need a strategy to make sure the new code is properly tested and to get that implemented. I mostly wanted to just prove out in the code that this was possible~~ The new jobs routine is fully tested.

**EDIT 02/20:** I added white box testing that exercises the new job branches. I'm not too concerned with this now given a green `make ci` + `make check`.

~~**TODO**: I still need to implement the `Sync` variants of the apb actions. I'm seeing them simply call the async versions and block while watching the channel. The caller will know no better, and we're using the same codepaths.~~ This is required, but the domain of another PR.

**Which issue this PR fixes (This will close that issue when PR gets merged)**
fixes #771

eriknelson commented on pkg/apb/executor.go:
```diff
@@ -32,11 +32,74 @@ import (
+	return e.extractedCredentials
+}
+
+func (e *Executor) start() {
```
They're really lifecycle hooks that need to be called from the action methods to alert the `Executor` to the current state of what's going on (and it abstracts the fact that messages are being sent). I originally named `updateDescriptor` to something like `sendDescriptorMessage`, but I felt like the latter was a less abstract name. I want to call high level hooks to alert the executor to triggers in the cycle without saying, "hey send this message", because it's better if the actions don't even care about that. They just say "hey, this thing happened, do what you would like with this info".

eriknelson replied:
I'll also say that I tried to avoid imposing additional requirements on callers since I expect these to turn into the primary methods that people will use with `libapb`. The worst of it is to ask folks to instantiate an `Executor` with `NewExecutor`; I've personally seen us on this path for a while, so I'm happy to finally get this down on paper, so to speak.

eriknelson replied:
Do we think there is a better name for `Executor`? I'm still hesitant to name these something like "sendX" or other related messaging concept. They're internal lifecycle hooks intended to be called by the internal APB code, and do more than send a message over the buffer (not much today, but tomorrow they might).

> also would give an opportunity for this to be an interface that could be overridden for tests or other uses if we exposed it via a constructor.

Definitely interested in this since decent test coverage is the biggest thing missing right now. I don't think callers should have to worry about injecting that via ctor though. A sensible default that can get overriden by test code I think makes sense, given this is an internal detail.

maleck13 replied:
> Definitely interested in this since decent test coverage is the biggest thing missing right now. I don't think callers should have to worry about injecting that via ctor though. A sensible default that can get overriden by test code I think makes sense, given this is an internal detail.

Yep agree no need to expose as long as it can be overridden in test code

eriknelson commented on pkg/apb/bind.go:
```diff
@@ -37,41 +37,55 @@ func Bind(instance *ServiceInstance,
+		metrics.ActionStarted("bind")
+		executionContext, err := e.executeApb(
+			"bind", instance.Spec, instance.Context, parameters)
+		defer runtime.Provider.DestroySandbox(
```
I'm not clear on the best practice with defers. If I'm doing something and I know I'll need to call the relevant "finish" function at the end of it, it makes sense to me to defer the call immediately after, assuming the mechanics of defer apply. A good example of this is the metrics hooks in the `apbJob`, I just partner the Start with a defered finish right way. I'll take a closer look at this, thanks for raising it.

eriknelson replied:
After looking at this, I think it doesn't actually matter in practice? Regardless of where the defer call is made, it will execute at the conclusion of the function. Optically, it might make sense to be made elsewhere, though.

jmrodri replied:
It doesn't matter, what I was saying was tying statements together.

```golang
    go func() {
        e.start()
        metrics.ActionStarted("bind")

        defer runtime.Provider.DestroySandbox(
            executionContext.PodName,
            executionContext.Namespace,
            executionContext.Targets,
            clusterConfig.Namespace,
            clusterConfig.KeepNamespace,
            clusterConfig.KeepNamespaceOnError,
        )

        executionContext, err := e.executeApb("bind", instance.Spec, instance.Context, parameters)
        if err != nil {
            log.Errorf("Problem executing apb [%s] bind", executionContext.PodName)
            e.finishWithError(err)
            return
        }
```

So the `err != nil` follows the executeApb so it is obvious that it is related and highly unlikely that someone will come and add another line of code that set err. As it stands now the defer statement separates the executeApb and the err check. If someone comes in later they could accidentally set err and we'll be eating an err.

eriknelson replied:
The issue is that `executionContext` isn't in scope until the `executeApb` call finished. I'll update to keep the err check tied closely.

shawn-hurley commented on pkg/apb/executor.go:
```diff
@@ -32,11 +32,74 @@ import (
-	context *Context,
-	p *Parameters) (ExecutionContext, error) {
+// Executor - Main struct used for running APBs. Manages their lifecycle.
+type Executor struct {
```
I would like to make it more clear that a single Executor for a single APB run. At least that is how I am reading this code am I wrong?

eriknelson replied:
Yeah that's exactly correct. What do you think the best way is to convey that?

shawn-hurley replied:
I would say documentation for the end user. Code examples will help too. I think that nothing from the code will make it more obvious. At least I can't think of anything at the moment.

maleck13 commented on pkg/broker/jobs.go:
```diff
@@ -0,0 +1,242 @@
+	specID                 string
+	bindingID              *string
+	method                 apb.JobMethod
+	metricsJobStartHook    metricsHookFn
```
So potentially as part of the work I have outlined under the proposal for the changes to the work engine these would become StateObservers / Subscribers and not needed here

maleck13 commented on pkg/apb/provision_or_update.go:
```diff
@@ -35,10 +35,8 @@ const (
-	stateUpdates chan<- JobState) (string, *ExtractedCredentials, error) {
-
+func (e *Executor) provisionOrUpdate(method executionMethod, instance *ServiceInstance) {
+	e.start()
```
Also see comment below around the start method

eriknelson commented on pkg/apb/executor.go:
```diff
@@ -32,11 +32,74 @@ import (
+	status.State = StateFailed
+	e.lastStatus = status
+	e.statusChan <- status
+	close(e.statusChan)
```
Very good point. Reading a bit, it looks like there isn't a safe way to check if a channel has been closed without reading from it, which feels ugly and indirect to me. Alternatively, we can lock the sensitive code and nil it out, then check if it's still present before closing.

eriknelson replied:
Added some locks.

eriknelson commented on pkg/apb/executor.go:
```diff
@@ -32,11 +32,74 @@ import (
+}
+
+// GetPodName - Returns the name of the pod running the APB
+func (e *Executor) GetPodName() string {
```
Updated!

shawn-hurley commented on pkg/apb/executor.go:
```diff
@@ -32,11 +33,117 @@ import (
+}
+
+// Executor - Composite executor interface.
+type Executor interface {
```
👍

shawn-hurley commented on pkg/apb/types.go:
```diff
@@ -380,17 +389,17 @@ type ExecutionContext struct {
-// Provisioner defines a function that knows how to provision an apb
-type Provisioner func(si *ServiceInstance, statusUpdate chan<- JobState) (string, *ExtractedCredentials, error)
+// Provisioner - a function that knows how to provision an apb
+type Provisioner func(si *ServiceInstance) (<-chan StatusMessage, error)
```
I think we can delete these interfaces with this PR?

eriknelson replied:
I've actually been going back and forth on this. I deleted them and then force pushed them back in. I'm trying to test asb_jobs, and the `runFn` should render these moot, so I'll probably pull them again.

jmrodri replied:
I don't see Provisioner used anywhere.

shawn-hurley commented:
I would also like us to make sure that we are not Exposing methods and types that we may not want to. I am thinking specifically of the ExtractCredentials functions.

jmrodri commented on pkg/apb/executor.go:
```diff
@@ -32,11 +33,117 @@ import (
+	Bind(instance *ServiceInstance, parameters *Parameters) <-chan StatusMessage
+	Unbind(instance *ServiceInstance, parameters *Parameters) <-chan StatusMessage
+	Update(instance *ServiceInstance) <-chan StatusMessage
+}
```
Will there be an `ExecutorSync`?

eriknelson commented:
@shawn-hurley Are you taking issue with the `Executor.ExtractedCredentials`? I went back and forth on this one... originally I did not have an accessor because it's a pointer type and could easily be mutated by a caller, but I added the accessor method for consistency because it felt silly to have the rest of the fields retrieved by method and not this one.

I think you should be able to get `ExtractedCredentials` from the `Executor` somehow.

shawn-hurley commented:
No we are exposing how we pull creds from the pod in the ext_creds.go I
think.

If it’s out of scope just tell me to go away until your done :)

On Mon, Feb 19, 2018 at 4:52 PM Erik Nelson <notifications@github.com>
wrote:

> @shawn-hurley <https://github.com/shawn-hurley> Are you taking issue with
> the Executor.ExtractedCredentials? I went back and forth on this one...
> originally I did not have an accessor because it's a pointer type and could
> easily be mutated by a caller, but I added the accessor method for
> consistency because it felt silly to have the rest of the fields retrieved
> by method and not this one.
>
> I think you should be able to get ExtractedCredentials from the Executor
> somehow.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/ansible-service-broker/pull/773#issuecomment-366811302>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABza922jf5j9eqVs8QXpo4_OJ4Qg3J8Xks5tWe0BgaJpZM4SJ47R>
> .
>

eriknelson commented:
@shawn-hurley can you elaborate? This is all good discussion; we're going to have to figure out how to split up the changes so this refactoring doesn't apply to the whole world. I would really like to keep PRs small and focused, and I know I've expressed that to you. I'm completely violating that here.

jmrodri commented on pkg/apb/bind.go:
```diff
@@ -37,41 +37,55 @@ func Bind(instance *ServiceInstance,
+			executionContext.Namespace,
+			instance.Spec.Runtime,
+		)
+		e.extractedCredentials = creds
```
What if there is an err with `ExtractCredentials`? Do we want to set `e.extractedCredentials` before checking the err value? If this is a conscience decision then okay, just wanted to make sure we're not setting something invalid.

eriknelson replied:
It was actually originally written this way, I thought about it and the expectation is that creds will be nil if an error is returned from `ExtractCredentials`. Your comment is probably safer, however. Will update.

jmrodri commented on pkg/apb/deprovision.go:
```diff
@@ -34,42 +34,49 @@ func Deprovision(instance *ServiceInstance, stateUpdates chan<- JobState) (strin
+		if err != nil {
+			log.Errorf("Problem executing apb [%s] deprovision", executionContext.PodName)
+			e.finishWithError(err)
+			return
```
same comment from bind.go regarding defer separating the err check.

jmrodri commented on pkg/apb/provision.go:
```diff
@@ -30,9 +27,7 @@ func Provision(
-	return provisionOrUpdate(executionMethodProvision, instance, stateUpdates)
+	go e.provisionOrUpdate(executionMethodProvision, instance)
+
+	return e.statusChan
```
:+1: provision.go looks okay

jmrodri commented on pkg/apb/update.go:
```diff
@@ -27,9 +27,7 @@ func Update(instance *ServiceInstance, statusUpdate chan<- JobState) (string, *E
-	// provisionOrUpdate as an implementation detail.
-	return provisionOrUpdate(executionMethodUpdate, instance, statusUpdate)
+	go e.provisionOrUpdate(executionMethodUpdate, instance)
+
```
update.go is okay

jmrodri commented on pkg/apb/provision_or_update.go:
```diff
@@ -76,31 +78,39 @@ func provisionOrUpdate(method executionMethod,
 		instance.Spec.Runtime,
 	)
+	e.extractedCredentials = creds
+
```
same comment as bind, are we sure we want to set this in case of an err. If so, okay.

jmrodri commented on pkg/apb/types.go:
```diff
@@ -139,6 +139,13 @@ type ExtractedCredentials struct {
+	Description string
+	Error       error
+}
+
```
+1 to `StatusMessage`

jmrodri commented on pkg/apb/unbind.go:
```diff
@@ -34,31 +37,42 @@ func Unbind(instance *ServiceInstance, parameters *Parameters, stateUpdates chan
-	return nil
+		e.finishWithSuccess()
+	}()
+
```
unbind.go looks okay

jmrodri commented on pkg/apb/watch_pod.go:
```diff
@@ -34,7 +34,12 @@ var (
+
+func watchPod(
+	podName string, namespace string, podClient v1.PodInterface,
+	updateDescription updateDescriptionFn,
```
watch_pod changes :+1:

jmrodri commented on pkg/broker/broker.go:
```diff
@@ -1241,7 +1242,7 @@ func (a AnsibleBroker) Update(instanceUUID uuid.UUID, req *UpdateRequest, async
 	log.Debugf("PreviousValues: [ %+v ]", req.PreviousValues)
 	log.Debugf("ServiceInstance Parameters: [%v]", *si.Parameters)
-	ujob := NewUpdateJob(si, apb.Update)
+	ujob := &UpdateJob{si}
```
I'm okay with using the job structs directly :+1:

jmrodri commented on pkg/broker/jobs.go:
```diff
@@ -0,0 +1,248 @@
+			return e.Update(j.serviceInstance)
+		},
+	}
+	job.Run(token, msgBuffer)
```
I looked over the entire jobs.go. :exclamation:  looks great. I like it.

jmrodri commented on pkg/metrics/metrics.go:
```diff
@@ -161,14 +161,14 @@ func DeprovisionJobStarted() {
-func BindingJobStarted() {
+// BindJobStarted - Add a provision job to the counter.
+func BindJobStarted() {
 	defer recoverMetricPanic()
```
okay, this is fine.

jmrodri commented on pkg/apb/executor.go:
```diff
@@ -32,11 +33,117 @@ import (
+	status := e.lastStatus
+	status.State = StateInProgress
+	e.lastStatus = status
+	e.statusChan <- status
```
I'm confused by this, we copy the lastStatus into status, set its state, then set it back on lastStatus. Why isn't this just:

```golang
func (e *executor) start() {
    log.Debug("executor::start")
    e.lastStatus.State = StateInProgress
    e.statusChan <- e.lastStatus
}
```

eriknelson replied:
I think I was on autopilot here and it felt like a redux action, which is of course totally irrelevant here. I'll update :smile:

shawn-hurley commented:
@eriknelson re: extracted credentials

This is what I was thinking about: https://github.com/openshift/ansible-service-broker/blob/master/pkg/apb/ext_creds.go#L48

I agree that this should be a follow-on, your right that this is something that can be done separately. I mentioned it here because #771  I thought was referring to the `apb` package and not just the actions. Sorry if this is off topic.

eriknelson commented:
@shawn-hurley I actually had the same thought and had marked it private, but the broker actually calls it in broker.go as part of recovery.

shawn-hurley commented:
I didn't realize that we may want to move recover out of the broker then. I think that may require a new issue to track and discuss, thoughts @jmrodri @rthallisey @eriknelson?

eriknelson commented:
@shawn-hurley IMO recover is firmly a concept in the broker domain. I'm not sure where else it would belong aside from there.

eriknelson commented:
@shawn-hurley @maleck13 @rthallisey Think this should be considered complete at this point. I'm planning on submitting another PR with the sync work.

eriknelson commented:
This utterly ballooned to almost unreviewable after I added the vendor stuff. If you look at the individual commits prior to that, I added `jobs.go` tests and in an effort to test it completely, I needed an `Executor` mock. I used [mockery](https://github.com/vektra/mockery) to generate `mock_Executor.go`, which was enormously helpful. It requires [testify](https://github.com/stretchr/testify), which I had to add to the vendor deps.

Gotta say, it was crazy easy to integrate and incredibly useful. I would love for us to continue with that pattern.

shawn-hurley commented on pkg/apb/mock_Executor.go:
```diff
@@ -0,0 +1,134 @@
+// Code generated by mockery v1.0.0
```
I am wondering if we should have a mock package for this? 

Just makes it more clear in your import that you're importing a mock?

eriknelson replied:
I started thinking about this as well, but it implies each pkg would have a corresponding child mock package? Does that make sense?

maleck13 replied:
what if you were to create a top level mock package that contained shared mocks instead of a mock package in each individual package?

eriknelson replied:
I don't think we want to be mixing mocks for a bunch of different packages into one package.

eriknelson commented:
Thank you for the reviews everyone! Merging since there is a bunch of work dependent on this, and it's green.
